ENT-3418: Freebsd packages

craigcomstock opened the pull request.

Lex-2008 commented on deps-packaging/apache/freebsd/cfbuild-apache-MANIFEST:
```diff
@@ -0,0 +1,457 @@
+    }
+}
+files {
+    "/var/cfengine/httpd/modules/mod_authn_file.so" "6aee1fdf0bc515d9e468aa0f57bd389d3816dcc6f8484731632de0dfbc902908"
```
when we update version of Apache, all these shas have to be updated, too. Is it possible to generate them dynamically, like for cfengine?

BasementTrix replied:
The gen_filelist() function in deps-packaging/pkg-build-freebsd recalculates the sha256 values.  Just delete the files{} section from the deps-packaging/*/freebsd/*-MANIFEST file before rebuilding the package.  In hindsight, I should have included logic to keep pkg-build-freebsd from adding a new files{} section to a MANIFEST that already has one.

craigcomstock replied:
@BasementTrix so... how should we resolve this? If I understand correctly we would need to regenerate the files{} section whenever we changed versions of a dep right?

craigcomstock replied:
@BasementTrix any ideas on what if anything we should do to make changing versions of deps easy? What would be the process? delete all the files {} sections in *-MANIFEST files? Could we make a helper script and add to docs so if you update deps-packaging/apache/distfiles for example, please run update-freebsd-manifest-checksums (or better named script) which would take care of deleting and regenerating files {}?

Lex-2008 commented on deps-packaging/apache/freebsd/cfbuild-apache-MANIFEST:
```diff
@@ -0,0 +1,457 @@
+    APACHE20
+]
+www = "http://httpd.apache.org/"
+maintainer = " Trix Farrar <trix@basement.net>"
```
I don't have a strong opinion here, but our debian/control files say "Maintainer: CFEngine Packager <packager@cfengine.com>".

On the other side, since these packages ae mostly "for internal use", I believe we don't mind having contributor's name here. Especially while we don't provide commercial support for this platform

BasementTrix replied:
It's my hope that they will eventually make it to general use, but wasn't sure what to put since I'm not packager@cfengine.com.  Feel free to change it.

Lex-2008 commented on deps-packaging/help2man/distfiles:
```diff
@@ -0,0 +1 @@
+7424901e6eca8af372d2ad62be5cf659 help2man-1.47.6.tar.xz
```
is help2man required? AFAIK, in order to be built, it should be added to the DEPS variable in compile-options script above. Or am I missing something?

Also, we usually try to use gzipped archives, since gzip is available on all platforms. But if this dependency is built only on FreeBSD, this probably is fine.

Also, don't you need a manifest file for it?

BasementTrix replied:
It's a leftover.  That whole directory can be dropped.  Same with indexinfo.

nickanderson replied:
Are you deleting that or waiting for us @BasementTrix?

Lex-2008 commented on deps-packaging/lcov/freebsd/build:
```diff
@@ -0,0 +1,34 @@
+		NEW_LANG='/usr/bin/env bash' ;;
+	esac
+	if [ -n "${NEW_LANG}" ] ; then
+	    /usr/bin/sed -i '' -e "s,${CURR_LANG}.*,${NEW_LANG}," "$FILENAME"
```
if we want to fix it in shebang **only**, can we ask `s` command to operate only on first line? Just add `1` before `s` :-)

Also, I'm not sure if you need an empty string and `-e` arguments here... Can't you just say `sed -i "<sed_expr>" "<filename>"`?

craigcomstock replied:
this one should be resolved, @Lex-2008 can you close this comment thread? (I don't recall if you can or not) Regardless, this one won't block merging.

Lex-2008 commented on deps-packaging/lcov/freebsd/build:
```diff
@@ -0,0 +1,34 @@
+	CURR_LANG=`head -1 ${FILENAME} | sed -e 's/^#!//; s/ .*//;'`
+	NEW_LANG=''
+	case "${CURR_LANG}" in
+	    /usr/bin/perl)
```
I would maybe err on "too broad" side and write `*perl*` here - in order to match on all shebangs including "perl", be it in `/bin`, `/usr/bin`, `/usr/local/bin`, etc, etc... Also, then you can remove sed command three lines above :-)

TrixFarrar-GEODIS replied:
Removing the sed invocations doesn't cover the case where the shebang line is '#!/usr/bin/perl -w'  First sed pulls out '/usr/bin/perl' for $CURR_LANG, second replaces (only) $CURR_LANG with $NEW_LANG.  For Perl, $NEW_LANG covers the case where Perl used to be /usr/bin/perl, (and may be linked) but is currently /usr/bin/perl.  Perl used to be in the base, the lang/perl5 port used to replace /usr/bin/perl with a link to /usr/local/bin/perl bash has never been part of the FreeBSD base, so it will always be in /usr/local/bin/bash.

TrixFarrar-GEODIS replied:
In my experience, '#!/usr/bin/perl -w' is more common than '#!/bin/bash -e' -- current project notwithstanding.

craigcomstock commented:
@BasementTrix can you take a look at Aleksei's comments? Thanks!

BasementTrix commented:
On Thu, Aug 02, 2018 at 09:55:10AM -0700, Aleksei Shpakovskii wrote:
> +      MFEST="${BASEDIR}/${PKG}/${PKG}-MANIFEST"
> 
> I would prefer using longer variable names. M-Fest sounds too much like a selebration :-)

But, it _is_ a celebration.  :-)  Nothing wrong with a global find/replace s/{MFEST}/{MANIFEST}/g,
I suppose.

> +      # Shuffle some files about for Community Edition
> +      if [ "${PROJECT}" = 'community' ]
> +      then
> +          /bin/mkdir -p ${CF_DIR}/share/CoreBase
> 
> can we use `${COREBASE}` var here, too?

I guess you could.  I'd have to look at the RPM definition to see how it's defined
and duplicate it.  Shouldn't be a big deal.

> +files {
> +    "/var/cfengine/httpd/modules/mod_authn_file.so" "6aee1fdf0bc515d9e468aa0f57bd389d3816dcc6f8484731632de0dfbc902908"
> 
> when we update version of Apache, all these shas have to be updated, too. Is it possible to generate them dynamically, like for cfengine?

If memory serves, I think you can delete out the checksums and deps-packaging/pkg-build-freebsd will
recalculate them.

> +www = "http://httpd.apache.org/"
> +maintainer = " Trix Farrar <trix@basement.net>"
> 
> I don't have a strong opinion here, but our debian/control files say "Maintainer: CFEngine Packager <packager@cfengine.com>".
> 
> On the other side, since these packages ae mostly "for internal use", I believe we don't mind having contributor's name here. Especially while we don't provide commercial support for this platform
> 

I can understand that -- but I'm not packager@cfengine.com, and I had to put something.  :-)

> > @@ -0,0 +1 @@
> +7424901e6eca8af372d2ad62be5cf659 help2man-1.47.6.tar.xz
> 
> is help2man required? AFAIK, in order to be built, it should be added to the DEPS variable in compile-options script above. Or am I missing something?
>

It's a leftover and can be removed.  At one point, I considered building everything, just for grins.
I got over it.

> > +PREFIX=${BUILDPREFIX:-/var/cfengine}
> +
> +fix_shebang()
> +{
> +    FILENAME=$1
> +    if [ -r "${FILENAME}" ] ; then
> +	CURR_LANG=`head -1 ${FILENAME} | sed -e 's/^#!//; s/ .*//;'`
> +	NEW_LANG=''
> +	case "${CURR_LANG}" in
> +	    /usr/bin/perl)
> +		NEW_LANG='/usr/bin/env perl' ;;
> +	    /bin/bash)
> +		NEW_LANG='/usr/bin/env bash' ;;
> +	esac
> +	if [ -n "${NEW_LANG}" ] ; then
> +	    /usr/bin/sed -i '' -e "s,${CURR_LANG}.*,${NEW_LANG}," "$FILENAME"
> 
> if we want to fix it in shebang **only**, can we ask `s` command to operate only on first line? Just add `1` before `s` :-)

That's fair.  Sloppy on my part.

> 
> Also, I'm not sure if you need an empty string and `-e` arguments here... Can't you just say `sed -i "<sed_expr>" "<filename>"`?

No, you can't.  GNU sed will assume a blank backup suffix (-i), but BSD sed does not.
BSD sed is in the base, GNU sed isn't.  I went with BSD syntax.

> 
> > @@ -0,0 +1,34 @@
> +#!/bin/sh -ex
> +
> +PREFIX=${BUILDPREFIX:-/var/cfengine}
> +
> +fix_shebang()
> +{
> +    FILENAME=$1
> +    if [ -r "${FILENAME}" ] ; then
> +	CURR_LANG=`head -1 ${FILENAME} | sed -e 's/^#!//; s/ .*//;'`
> +	NEW_LANG=''
> +	case "${CURR_LANG}" in
> +	    /usr/bin/perl)
> 
> I would maybe err on "too broad" side and write `*perl*` here - in order to match on all shebangs including "perl", be it in `/bin`, `/usr/bin`, `/usr/local/bin`, etc, etc... Also, then you can remove sed command three lines above :-)
> 

You're probably right.  The "standard" location has, to my memory, always been '/usr/bin/perl',
but you know standards.  On FreeBSD, perl is usually /usr/local/bin/perl and bash is /usr/local/bin/bash
-- which is what lead to the fix_shebang() hack to begin with.

-- 
John D. "Trix" Farrar            __\\|//__               Basement.NET
trix@basement.net                (` o-o ')   http://www.basement.net/
--------------------------------ooO-(_)-Ooo--------------------------
 GPG Key Fprint: 39D0 BCD2 9C9F B7EF B125  D419 61E8 8648 16F2 1814

Lex-2008 commented:
Indeed, sadly, Github doesn't split email replies into individual comment threads.

Regarding `${COREBASE}` my point was that it's already used 3 lines below :-)

Regarding SHAs - my quick googling yesterday didn't suggest that we can just skip them... But worth trying!

Also, thanks for enlighting me about GNU vs BSD sed! Indeed, I grew up in a GNU land where backup suffix is optional :-)

BasementTrix commented on build-scripts/package:
```diff
@@ -239,45 +239,134 @@ case "$PACKAGING" in
+      # Shuffle some files about for Community Edition
+      if [ "${PROJECT}" = 'community' ]
+      then
+          /bin/mkdir -p ${CF_DIR}/share/CoreBase
```
You're probably correct.

nickanderson replied:
@BasementTrix are you going to submit a change to variable for core base or are you waiting for us?

BasementTrix commented:
There is a better version of packaging/common/freebsd/cfengine3.rc
[cfengine3.rc.gz](https://github.com/cfengine/buildscripts/files/2256935/cfengine3.rc.gz)
 
(Obviously, I'm not a github maven)

BasementTrix commented:
On Fri, Aug 03, 2018 at 02:02:55AM -0700, Aleksei Shpakovskii wrote:
> Regarding SHAs - my quick googling yesterday didn't suggest that we can just skip them... But worth trying!

You're right.  pkg(8) will not let you skip the checksums, but the
deps-packaging/pkg-build-freebsd script will.  :-)

-- 
John D. "Trix" Farrar            __\\|//__               Basement.NET
trix@basement.net                (` o-o ')   http://www.basement.net/
--------------------------------ooO-(_)-Ooo--------------------------
 GPG Key Fprint: 39D0 BCD2 9C9F B7EF B125  D419 61E8 8648 16F2 1814

nickanderson commented on build-scripts/compile-options:
```diff
@@ -91,7 +92,7 @@ esac
 # iconv is needed for libxml2 on some platforms
 case "$OS_FAMILY" in
-   aix|solaris)  var_append DEPS "libiconv" ;;
+   aix|solaris|freebsd)  var_append DEPS "libiconv" ;;
```
It does look that way.

BasementTrix commented:
On Thu, Aug 16, 2018 at 01:47:44PM -0700, Nick Anderson wrote:
> @BasementTrix are you going to submit a change to variable for core base or are you waiting for us? 

Waiting for Craig.  I submitted a patch file, since my buildscripts local
copy is subversion, I wasn't sure I could do more than that.  That's also
why I submitted the improved copy of cfengine3.rc as a file attached
to this thread rather than a git commit.

-- 
John D. "Trix" Farrar            __\\|//__               Basement.NET
trix@basement.net                (` o-o ')   http://www.basement.net/
--------------------------------ooO-(_)-Ooo--------------------------
 GPG Key Fprint: 39D0 BCD2 9C9F B7EF B125  D419 61E8 8648 16F2 1814

nickanderson commented:
I see. I believe Craig will Mook tomorrow when he is back. If you mail the patch to me I can try to integrate it.

craigcomstock commented:
@BasementTrix I have added a commit with what I think are the needed revisions so far based on the conversation here. Let me know if I got it right. We also still need to write-up detailed build instructions. I will edit my notes into a usable doc now when I test out these most recent changes.

Thanks!

BasementTrix commented:
@craigcomstock exactly the way I would have done it.

As for the build steps:

export PROJECT=nova
export BUILD_TYPE=DEBUG
export ESCAPETEST=yes
export EXPLICIT_ROLE=agent
export EXPLICIT_VERSION='3.12.0'
cd ~/CFEngine-build
sudo pkg install -qy  devel/autoconf devel/automake devel/libtool
./buildscripts/build-scripts/autogen
sudo pkg delete  -x autoconf automake m4 libtool
sudo pkg install -qy devel/gmake lang/gcc6 security/sudo textproc/expat2 devel/git
./buildscripts/build-scripts/build-environment-check
./buildscripts/build-scripts/install-dependencies
./buildscripts/build-scripts/test
./buildscripts/build-scripts/configure
./buildscripts/build-scripts/compile
sudo pkg delete -x cfbuild-.\*-devel
./buildscripts/build-scripts/package

I've been told that this is (more or less) how a Jenkins build runs.  I do not have a working Jenkins installs running.

lang/gcc7 hit the Ports Collection last week.  I haven't tested that, but see no reason why it wouldn't work.
In the meantime, lang/gcc6 is still available and works fine.

craigcomstock commented:
@BasementTrix I went through my notes and looked at yours as well and came up with the README.freebsd file I just added. Can you take a look and comment?

craigcomstock commented:
Interestingly when I do ```sudo pkg install lang/python27``` I don't get a "python" in my path and I notice our determine-version.py script doesn't work as it's shebang is ```#!/usr/bin/env python```. So it seems like ```sudo pkg install -qy lang/python``` would be best.

TrixFarrar-GEODIS commented:
If it works for you, go for it.  :-)

I didn't take into account that lang/python is a meta-port that symlinks the "default python" to /usr/local/bin/python.

Currently, PYTHON_DEFAULT=2.7.  I'll let you know if/when it gets changed.

Trix Farrar
Senior Systems Administrator
GEODIS | Americas | Contract Logistics

7101 Executive Center Drive, Suite 333 – Brentwood, TN 37027 - US
Tel. : +1 615 401 6431 - Mob. : +1 615 934 8134
Trix.Farrar@geodis.com | geodis.com

________________________________________
From: craigcomstock <notifications@github.com>
Sent: Tuesday, August 28, 2018 3:46:02 PM
To: cfengine/buildscripts
Cc: Farrar, Trix; Comment
Subject: Re: [cfengine/buildscripts] ENT-3418: Freebsd packages (#422)

Interestingly when I do sudo pkg install lang/python27 I don't get a "python" in my path and I notice our determine-version.py script doesn't work as it's shebang is #!/usr/bin/env python. So it seems like sudo pkg install -qy lang/python would be best.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_cfengine_buildscripts_pull_422-23issuecomment-2D416734512&d=DwMFaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=JT6udsA09PmBNCuWul_2zibQSVA4uMQBaVXJ7G_MTPw&s=CAEv14SuUxcu2AA3qFTm7OBQ2xU7LgrD4RINNq-zLKY&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AfiJEeRCb8N9VlD-2De5A3J8G4HJAfraSpks5uVawKgaJpZM4Vseiv&d=DwMFaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=JT6udsA09PmBNCuWul_2zibQSVA4uMQBaVXJ7G_MTPw&s=w0lkdrxE8RJwVUqZcUtdYPI6RnAXEDh2SSnDqaiUR4k&e=>.

nickanderson commented:
It's been a couple days since we have seen anything here. What's the stauts?

TrixFarrar-GEODIS commented:
The hashes in package manifests are there to make pkg(8) happy and are recalculated each time the build is run.

The CRCs for downloaded files have to be generated when new versions come out.

Trix Farrar
Senior Systems Administrator
GEODIS | Americas | Contract Logistics

7101 Executive Center Drive, Suite 333 – Brentwood, TN 37027 - US
Tel. : +1 615 401 6431 - Mob. : +1 615 934 8134
Trix.Farrar@geodis.com | geodis.com

________________________________________
From: Nick Anderson <notifications@github.com>
Sent: Friday, January 11, 2019 2:12:18 PM
To: cfengine/buildscripts
Cc: Farrar, Trix; Comment
Subject: Re: [cfengine/buildscripts] ENT-3418: Freebsd packages (#422)

It's been a couple days since we have seen anything here. What's the stauts?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_cfengine_buildscripts_pull_422-23issuecomment-2D453643005&d=DwMCaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=H8KuIHzthhXbkYNDjj2fSqh1_urxIUWERTAhgcKzzzc&s=CxIAsqOK3lWSR4S_7-vgI6T3owsWJRqYPkD-1rVALx8&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AfiJEXAJex-5F-2Dica06IvPNEJRV853V4Unks5vCPAigaJpZM4Vseiv&d=DwMCaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=H8KuIHzthhXbkYNDjj2fSqh1_urxIUWERTAhgcKzzzc&s=xTGnAI71dqhk_tEJPTaZIkQjBiUBPwzGxLoOrrul5s8&e=>.

craigcomstock commented:
@BasementTrix were do you think we stand on this PR? If you think it is ready I think we could spend a little time vetting it and merging it. I was about to squash commits a little while ago so probably pretty much ready to go right?

TrixFarrar-GEODIS commented:
I think it's ready, but am prepared to deal with issues that may become apparent.

Trix Farrar
Senior Systems Administrator
GEODIS | Americas | Contract Logistics

7101 Executive Center Drive, Suite 333 – Brentwood, TN 37027 - US
Tel. : +1 615 401 6431 - Mob. : +1 615 934 8134
Trix.Farrar@geodis.com | geodis.com

________________________________________
From: craigcomstock <notifications@github.com>
Sent: Friday, March 1, 2019 10:00:48 AM
To: cfengine/buildscripts
Cc: Farrar, Trix; Comment
Subject: Re: [cfengine/buildscripts] ENT-3418: Freebsd packages (#422)

@BasementTrix<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_BasementTrix&d=DwMCaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=xF0k6dgLfHLK3Z0bnTxOc423Kh2swYho-N0LvPfLkoc&s=mJBbz3pSKIWIgSWHH0GSO9L-unITkOQ2V75u8dMbpoo&e=> were do you think we stand on this PR? If you think it is ready I think we could spend a little time vetting it and merging it. I was about to squash commits a little while ago so probably pretty much ready to go right?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_cfengine_buildscripts_pull_422-23issuecomment-2D468714159&d=DwMCaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=xF0k6dgLfHLK3Z0bnTxOc423Kh2swYho-N0LvPfLkoc&s=yCbY_VgIkiNDJepX3T_SoQQ2GMt5AJwVg2XL-JM4HNc&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AfiJEUPSBZbP6qZtBcwbEsSBk3sdKGtnks5vSU6wgaJpZM4Vseiv&d=DwMCaQ&c=xpl-dddEK2E16jaE4CUn4A&r=f72KDIFrJ-_GSB2OeKtFhjjNbHzARR9iedyzy9sUgAU&m=xF0k6dgLfHLK3Z0bnTxOc423Kh2swYho-N0LvPfLkoc&s=yNvy9nIVTYsxXrTj_c7BvrqZdWRxez6I8XhERRf1RbE&e=>.

craigcomstock commented:
> I think it's ready, but am prepared to deal with issues that may become apparent. 

Thanks @TrixFarrar-GEODIS. I have squashed the commits and rebased from master. Will try a build and see where we get! Sorry it's taken so long to get this in. If you want to try this PR branch out that would be great too.

craigcomstock commented:
@TrixFarrar-GEODIS and/or @BasementTrix can you try this current PR out again? So far I have worked on community package and had some problem with libcurl. Thanks! :) I will stop actively working on this for a little while so if you can make some progress that would be great, otherwise we will come back to it when we can.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cfengine/buildscripts?pullRequest=422) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/cfengine/buildscripts?pullRequest=422) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: BasementTrix<br/>:x: craigcomstock<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cfengine/buildscripts?pullRequest=422) it.</sub>

craigcomstock commented:
Looking at the number of changes here and such, I'm wondering if it wouldn't be better to look at something like FPM? https://github.com/jordansissel/fpm The project is active and supports deb, rpm, solaris, freebsd, tar, osxpkg, archlinux. I may take some value add time and experiment with this strategy. @olehermanse @nickanderson @Lex-2008 @vpodzime, thoughts?

craigcomstock commented:
Given the large number of changes and lack of activity I don't think it makes sense to pursue this at this time.
